### PR TITLE
dev-lang/python: enable profile guided optimization

### PR DESCRIPTION
**Pull request by InBetweenNames**
This solves bug #615412.  Add a USE flag for PGO
that is off by default, but if enabled, will trigger the appropriate
make target to do a PGO build of the Python interpreter.

Ping @gentoo/python

Package-Manager: Portage-2.3.10, Repoman-2.3.3

### CODE REVIEW

**Comment by ThGravo** [comment]
Is there a specific reason why you don't use `--enable-optimizations`?

**Comment by InBetweenNames** [comment]
Yes--because that enables other optimization flags as well.  Gentoo policy is to not enable different CFLAGS with USE flags--PGO is one exception because it doesn't behave like LTO or O3--it requires multi stage compilation including instrumentation and training.  I've tested all of these ebuilds on my own machine and they build correctly here.

So this allows for PGO but also lets the user decide which other CFLAGS they'd like to use

**Comment by austin987** [comment]
IMO this is something that should be handled upstreamed. I.e. an --enable-pgo option that --enable optimizations also triggers, but less maintenance long term.

Do you know if/have you contacted upstream about that?

**Comment by InBetweenNames** [comment]
I agree--it actually is supported upstream through `profile-opt` however, which is well documented.  I imagine all a `--enable-pgo` would do is make `profile-opt` the default Makefile target.  `--enable-optimizations` does that and also forces `-O3`, which is of course against Gentoo policy--otherwise I'd just use that.  I think it would be hard to get such a patch pushed through to Python as the first response I'd get would probably be "why doesn't `make profile-opt`" work?  We already use that."

**Comment by InBetweenNames** [comment]
One thread from 2015 states that the Python binaries themselves are built with:

~~~
make profile-opt PROFILE_TASK="-m test.regrtest -w -uall,-audio -x test_gdb
test_multiprocessing"
~~~

Source: https://mail.python.org/pipermail/python-dev/2015-August/141328.html

It's a very strongly worded thread and I don't agree with a lot of the views there, but I did find that part interesting.

**Comment by candrews** [comment]
_bump_ what's the latest here?

**Comment by InBetweenNames** [comment]
I actually ended up providing ebuilds in my own overlay, `lto-overlay`, that add a USE flag for PGO with all Python versions.  I think it'd be great if we could get them supported upstream, as most, if not all binary distributions use PGO with their Python builds, and Gentoo in particular relies on Python so heavily.

Gentoo Developers: if there is anything I need to do in order to make that happen, just let me know.

**Comment by floppym** [comment]
Looking into this now.

Related bug: https://bugs.gentoo.org/615412

**Comment by floppym** [comment]
So this basically got stalled on my system due to a sandbox error when running the test suite to generate pgo data with Python 3.6.x. I'll see if I can work around it.

**Comment by InBetweenNames** [comment]
@floppym Have you tried the ebuilds from `lto-overlay`?  They are more up to date than the ones in this PR here.

**Comment by floppym** [comment]
Can you summarize the differences? I see you are overriding `PROFILE_TASK`, but I'm not sure if you are doing anything else.

Regarding `PROFILE_TASK`, it would be great if you could get that exclusion list pushed upstream (https://github.com/python/cpython).

**Comment by InBetweenNames** [comment]
Indeed, it's just the exclusion lists that are different.  They were modified to accommodate a bug report we had on `lto-overlay` where someone was getting loopback errors for the `asyncio` tests, and also the Python 3.4 exclusion list was modified as I got errors on that one that I didn't get with 3.5 and 3.6.  My lists came from the mailing list post linked earlier in this thread.

As for upstreaming the exclusion lists, I'm not sure if I'd have much luck in doing that, as Python developers have optimized their workflow for themselves and not for users.  That's why, for example, PGO is an opt-in rather than an opt-out feature, and it has to be accessed through that special makefile target.  There isn't even a configuration flag for PGO, just the generic `--enable-optimizations`, which carries different meanings depending on which version of Python you are using.

AFAIK, all binary distributions use some form of PGO with their Python distributions.  I have noticed Portage running a lot faster with PGO enabled myself.  Users of my overlay have reported similar results.

**Comment by mmokrejs** [comment]
Related pull request? https://github.com/gentoo/gentoo/pull/5482

**Comment by InBetweenNames** [comment]
That's fine, interested users can just get it from LTO-overlay.

**Comment by InBetweenNames** [comment]
Just so everyone knows, the PGO ebuilds in my repo are now parallelized.
